### PR TITLE
Issue 1088 leave game

### LINKED ISSUE #1088: Hook when a player leaves a match
Hi ! i can't find any way to execute some logic when a players leaves a match.

I'd like to handle cases where a player leave a match (by closing it's browser for example). Right now i'm using Pusher which send webhooks to my api when a user quits the application, which handle calling the lobby leave endpoint (https://boardgame.io/documentation/#/api/Lobby?id=leaving-a-match) using the appropriate player ID and player credentials.

I have a custom players state in G to handle players score, position on the board, etc...

After a player leaves the match, i'd like to remove his state and remove him from the turn order to prevent other players being blocked.

But when a player has left the match (when his "isConnected" attribute is set to false) i can't find a way to execute some logic to do that.

Should i use a server middleware on the leave route to update the game state ? and how ? Or Is there a better solution to achieve that ?

Can someone guide me into doing this ?

### PR DESCRIPTION
**Pull request by Rupesh-ark**
Closes #1088

  ## Summary

  This adds a permanent game-leave flow while keeping the existing lobby-slot behavior intact.

  - Keeps `/leave` as lobby metadata / slot cleanup only.
  - Adds `/leaveSlot` as the explicit slot-cleanup endpoint.
  - Adds `POST /games/:name/:id/leaveGame` for permanent game leave.
  - Adds `LobbyClient.leaveGame(...)`.
  - Adds `onPlayerLeave({ G, ctx, playerID, events })`.
  - Adds `events.removePlayer(playerID)` for game/server lifecycle use.
  - Runs game leave logic before clearing the lobby slot.
  - Keeps socket disconnect presence-only.

  ## Design decisions

  ### `/leave` remains slot-only

  `/leave` already behaves like lobby metadata cleanup. This PR keeps that behavior unchanged to avoid
  breaking existing apps that use it only to free a lobby seat.

  Permanent game leave now goes through `/leaveGame`, which is explicit and can run game-state
  lifecycle logic.

  ### Leaving a game is a lifecycle event, not only metadata cleanup

  A permanent leave can affect `G`, `ctx.playOrder`, `ctx.activePlayers`, undo/redo history, logs, and
  gameover state. Because of that, `/leaveGame` runs through the normal reducer/plugin/event lifecycle
  instead of mutating lobby metadata alone.

  ### Game authors own game-specific leave behavior

  The new `onPlayerLeave` hook lets game authors decide what leaving means for their game. The
  framework removes the player from engine turn/active-player bookkeeping, but game-specific cleanup
  belongs in the hook.

  This avoids a default behavior that could accidentally pause or break games whose rules need custom
  handling.

  ### `events.removePlayer` is server/lifecycle-only

  `events.removePlayer(playerID)` exists so leave logic can remove a non-current player safely. It is
  not exposed as a raw client-callable event. Client-originated private actions/events are rejected to
  prevent forged socket actions.

  ### `/leaveGame` runs lifecycle before slot cleanup

  The game state is updated first. Only after that succeeds does the server clear the lobby slot.

  This avoids losing the lobby/player metadata before the game has had a chance to authenticate and
  process the leave.

  ### Slot cleanup re-checks current metadata

  After lifecycle processing, `/leaveGame` fetches fresh metadata and checks credentials again before
  clearing the slot. This avoids clearing a slot that may have been rejoined/reclaimed during the
  leave flow.

  ### Game-over leave only clears the slot

  If the match is already over, there is no game lifecycle left to run. In that case, `/leaveGame`
  just clears the lobby slot instead of failing.

  ## Safety / edge cases

  This includes guards and regression coverage for the tricky cases around player removal:

  - Rejects forged private/internal client actions such as `PLAYER_LEAVE`.
  - Rejects raw socket attempts to call private events like `removePlayer`.
  - Validates URL game name against match metadata before player lookup.
  - Rejects non-string `playerID` values.
  - Re-authenticates against fresh metadata before clearing the lobby slot.
  - Cleans removed players from play order, active players, previous/queued active-player state.
  - Rebases undo/redo history so undo cannot restore pre-leave state.
  - Handles game-over leave requests by clearing the lobby slot without rerunning game lifecycle.

  ## Tests

  Added coverage across core, master, lobby client, and server API behavior.

  Verified locally:

  - `pnpm run lint`
  - `pnpm run ts`
  - targeted Jest tests
  - `pnpm run test:coverage`

  Full coverage run passed: 42 suites, 847 tests.

  ## Checklist

  - [x] Use a separate branch in your local repo, not `main`.
  - [x] Test coverage is 100% or there is a coverage story: new behavior is covered by targeted
  regression tests, and the full coverage suite passes.

### CODE REVIEW

File: src/server/api.ts [hunk 1]
```diff
@@ -15,6 +15,17 @@ import cors from '@koa/cors';
 import { createMatch, getFirstAvailablePlayerID, getNumPlayers } from './util';
 import type { Auth } from './auth';
 import type { Server, LobbyAPI, Game, StorageAPI } from '../types';
+import { Master } from '../master/master';
+import type { TransportAPI as MasterTransport } from '../master/master';
+
+interface MatchQueue {
+  add<T>(task: () => T | Promise<T>): Promise<T | Promise<T>>;
+}
```
[the remark this answers is not in the document]
Comment (Rupesh-ark) [reply src/server/api.ts:23]: Fixed in 98b8f6c — narrowed the task type to `add<T>(task: () => PromiseLike<T>): Promise<T>`. The queue is only ever given async tasks, and this keeps the interface structurally compatible with `PQueue.add` (typing it exactly as suggested made TypeScript infer `TaskResultType = T | Promise<T>` from p-queue's `Task` union and broke the `SocketIO` → `MatchTransport` assignment).

**Comment by Rupesh-ark** [comment]
Addressed the review feedback and the failing coverage check:

- **REDO fall-through in `LogEvent.svelte`** (Copilot's suppressed comment): added the missing `break` so REDO actions render as `redo` instead of falling through to `action.payload.type` (98b8f6c).
- **`MatchQueue.add` return type**: narrowed to `add<T>(task: () => PromiseLike<T>): Promise<T>` (98b8f6c, see inline reply for why the exact suggestion didn't typecheck).
- **coverage/coveralls**: the drop came from 24 uncovered lines and a set of uncovered branches in the new leave-lifecycle code. Added tests covering the reducer rejection paths, turn-order removal bookkeeping edge cases, `Master.onPlayerLeave` error/patch/gameover paths, leave-endpoint validation and transport queue wiring, log redaction of arg-less actions, and `SocketIO.createTransportAPI` (dcacd7e). The repo is back to 100% line and branch coverage locally.

File: src/server/api.ts [hunk 2]
```diff
@@ -82,18 +93,116 @@ export const configureRouter = ({
...
+      ctx.throw(404, 'Player ' + playerID + ' not found');
+    }
+    const isAuthorized = await auth.authenticateCredentials({
+      playerID,
+      credentials,
+      metadata,
+    });
+    if (!isAuthorized) {
+      ctx.throw(403, 'Invalid credentials ' + credentials);
+    }
```
[the remark this answers is not in the document]
Comment (Rupesh-ark) [reply src/server/api.ts:167]: Fixed in 0353e7b — the 403 now returns a generic `Invalid credentials` message. Also applied the same fix to the two pre-existing occurrences in the play-again and update-player endpoints so credentials are no longer reflected anywhere in api.ts, and updated the affected tests.

File: src/core/turn-order.ts [hunk 3]
```diff
@@ -19,17 +19,127 @@ import type {
...
+      if (currentPlayerPos === -1) {
+        playOrderPos =
+          ctx.playOrderPos > playOrder.length - 1 ? 0 : ctx.playOrderPos;
+        currentPlayer = playOrder[playOrderPos];
+      } else {
+        playOrderPos = currentPlayerPos;
+      }
+    }
+  }
+
```
[the remark this answers is not in the document]
Comment (Rupesh-ark) [reply src/core/turn-order.ts:87]: This state is reachable when an `onPlayerLeave` hook ends the game before the final player is removed. Fixed in fe7241e: `RemovePlayer` now sets `currentPlayer` to `''` when the removal empties `playOrder`, matching `InitTurnOrderState`, with tests asserting the normalized ctx at both the turn-order and reducer level.

**Review by devill — COMMENTED** [review]
Solid PR — clean endpoint split, good `onUpdate` allowlist, strong tests (42 suites / 879 green locally, ts + lint clean).

Approve with minor changes. No blockers — two small notes inline (one doc, one error-message nit).

**Review by devill — APPROVED** [review]

**Comment by devill** [comment]
@Rupesh-ark  Decided to merge as is. The tiny message leak can be fixed in a separate PR.

**Comment by Rupesh-ark** [comment]
@devill Imma raise another pr for that, thanks for the review :)
